File: go/ocr-numbers/ocr_numbers.go
package ocr

import (
	"strings"
)

func Recognize(input string) []string {
	separateLines := strings.Split(input, "\n")
	numbers := []string{}
	for i := 1; i < len(separateLines); i += 4 {
		number := ""
		for j := 0; j < len(separateLines[i]); j += 3 {
			digit := separateLines[i][j:j+3] + "\n" + separateLines[i+1][j:j+3] + "\n" + separateLines[i+2][j:j+3]
			number += recognizeDigit(digit)
		}
		numbers = append(numbers, number)
	}
	return numbers
}

func recognizeDigit(digit string) string {
	switch digit {
	case " _ \n| |\n|_|":
		return "0"
	case "   \n  |\n  |":
		return "1"
	case " _ \n _|\n|_ ":
		return "2"
	case " _ \n _|\n _|":
		return "3"
	case "   \n|_|\n  |":
		return "4"
	case " _ \n|_ \n _|":
		return "5"
	case " _ \n|_ \n|_|":
		return "6"
	case " _ \n  |\n  |":
		return "7"
	case " _ \n|_|\n|_|":
		return "8"
	case " _ \n|_|\n _|":
		return "9"
	default:
		return "?"
	}
}

File: elixir/ocr-numbers/lib/ocr_numbers.ex
defmodule OcrNumbers do
  @numbers %{{' _ ', '| |', '|_|', '   '} => "0", {'   ', '  |', '  |', '   '} => "1", {' _ ', ' _|', '|_ ', '   '} => "2", {' _ ', ' _|', ' _|', '   '} => "3", {'   ', '|_|', '  |', '   '} => "4", {' _ ', '|_ ', ' _|', '   '} => "5", {' _ ', '|_ ', '|_|', '   '} => "6", {' _ ', '  |', '  |', '   '} => "7", {' _ ', '|_|', '|_|', '   '} => "8", {' _ ', '|_|', ' _|', '   '} => "9"}

  @doc """
  Given a 3 x 4 grid of pipes, underscores, and spaces, determine which number is represented, or
  whether it is garbled.
  """
  @spec convert([String.t()]) :: {:ok, String.t()} | {:error, String.t()}
  def convert(input) do
    cond do
      rem(length(input), 4) != 0 -> {:error, "invalid line count"}
      Enum.any?(input, &(rem(String.length(&1), 3) != 0)) -> {:error, "invalid column count"}
      true -> {:ok, do_convert(input)}
    end
  end

  defp do_convert(input) do
    input
    |> Enum.chunk_every(4)
    |> Enum.map(&construct_line/1)
    |> Enum.join(",")
  end

  defp construct_line(input) do
    input
    |> Enum.map(&String.to_charlist/1)
    |> Enum.map(&Enum.chunk_every(&1, 3))
    |> Enum.zip()
    |> Enum.map(&(if (num = @numbers[&1]) != nil, do: num, else: "?"))
    |> Enum.join("")
  end
end
